Some updates to scheduling policies docs (#5911)
dask/distributed: docs/source/scheduling-policies.rst

Tasks that don't meet the root-ish criteria described above are selected as follows:

First, we identify the pool of viable workers:

1.  If the task has no dependencies and no restrictions, then we find the
    least-occupied worker.
2.  Otherwise, if a task has user-provided restrictions (for example it must
    run on a machine with a GPU) then we restrict the available pool of workers
    to just that set. Otherwise, we consider all workers.
3.  We restrict the above set to just workers that hold at least one dependency
    of the task.

From among this pool of workers, we then determine the worker where we think the task will start

> running the soonest, using :method:`Scheduler.worker_objective`. For each worker:

Reviewer: ```suggestion
running the soonest, using :meth:`Scheduler.worker_objective`. For each worker:
```

Reviewer: :method: describes a method, :meth: links one.
https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html

1.  We consider the estimated runtime of other tasks already queued on that worker.
    Then, we add how long it will take to transfer any dependencies to that worker that
    it doesn't already have, based on their size, in bytes, and the measured network

>     bandwith between workers. (This does *not* consider serialization time, nor whether
>     the data is spilled to disk versus in memory on the worker.) In practice, the

Reviewer: ```suggestion
    bandwith between workers. Note that this does *not* consider (de)serialization
    time, time to retrieve the data from disk if it was spilled, or potential differences
    between size in memory and serialized size. In practice, the
```
nor should only be used in a neither...nor construct

    queue-wait-time (known as *occupancy*) usually dominates, so data will usually be
    transferred to a different worker if it means the task can start any sooner.
2.  It's possible for ties to occur with the "start soonest" metric, though uncommon
    when all workers are busy. We break ties by choosing the worker that has the

>     fewest number of bytes of Dask data stored (including spilled data). Note that
>     this is not the same as :ref:`managed memory <memtypes>`, nor the total memory
>     used by the process.

Reviewer: ```suggestion
    fewest number of bytes of Dask data stored (including spilled data). Note that
    this is the same as :ref:`managed <memtypes>` plus :ref:`unmanaged <memtypes>`
    memory, not the :ref:`process <memtypes>` memory.
```

Reviewer: Sorry, typo - I meant to write plus spilled.
In the referenced high level document as well as in the GUI, "managed" memory maps to ``MemoryState.managed_in_memory``, not to ``MemoryState.managed`` which also includes spilled. This is a deliberate misalignment to reduce confusion in newer users.

Reviewer: ``MemoryState.managed_in_memory = min(WorkerState.nbytes, MemoryState.process)`` so in most cases they are the same.

Last in, first out
~~~~~~~~~~~~~~~~~~

When a worker finishes a task, the immediate dependencies of that task get top
priority.  This encourages a behavior of finishing ongoing work immediately

> before starting new work.  This often conflicts with the

Reviewer: ```suggestion
before starting new work (depth-first graph traversal). This often conflicts with the
```

first-come-first-served objective, but often results in shorter total runtimes

> and significantly reduced memory footprints.

Reviewer: short of avoiding disk spillage, why would it result in a shorter runtime?

    policy between computations.  All tasks from a previous call to compute
    have a higher priority than all tasks from a subsequent call to compute (or
    submit, persist, map, or any operation that generates futures).
3.  Whenever a task is ready to run (its dependencies, if any, are complete),
    the scheduler assigns it to a worker. When multiple tasks are ready at once,

>     the order in which they are scheduled is undefined, not based on priority.

Reviewer: this doesn't sound right. I understand that the scheduler should stop sending tasks to the workers once the task pressure on them is enough, and that there should still be a depth-first priority among ready tasks?

Author: Your intuition that this doesn't sound right is correct, because it's not good scheduling logic. However, it's right in that it's what actually happens today (second half is wrong actually, see below). 

> I understand that the scheduler should stop sending tasks to the workers once the task pressure on them is enough

If this happened, we wouldn't have root task overproduction. In fact, this is one of the solutions I suggested at the bottom of https://github.com/dask/distributed/issues/5223 (don't tell workers about extra tasks in the first place).

The criteria for a task to transition to `processing` is only that it's not waiting on any dependencies. (There's no throttling of the number of tasks that can be in `processing` at once; it can be far more than the workers actually have capacity for.) `transition_waiting_processing` is the only path for a task to get scheduled (besides stealing). `transition_waiting_processing` will only _not_ schedule the task if `decide_worker` returns None: https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/scheduler.py#L2737-L2739
`decide_worker` [only returns None](https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/scheduler.py#L2613-L2622) if there are no valid workers for the task, which is just a question of worker/host/resource restrictions. There's no logic checking whether the task pressure on the workers is too high.

Because this backpressure doesn't exist, the scheduler oversaturates workers with root tasks, and inevitable race conditions cause them to run too many of them, causing root task overproduction.

> there should still be a depth-first priority among ready tasks

```suggestion
    they are all submitted to workers, in priority order.
```

Reading more carefully, my original changes here were wrong, and we can actually say this ^. I initially said "the order in which they are scheduled is undefined" because `recommendations` aren't pre-sorted during `transitions`, so there's no mechanism guaranteeing or validating that they're scheduled in priority order. However:
1. When tasks initially get added via `update_graph`, the recommendations are added in priority order: https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/scheduler.py#L4875-L4877
2. When a task completes, its dependents are recommended in priority order: https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/scheduler.py#L8151-L8153

So in practice, tasks are assigned to workers in priority order. Indeed, https://github.com/dask/distributed/pull/4967 relies on this fact.

However, it's important to note the "all" here. Yes, `decide_worker` does end up being called on the tasks in priority order, but because every runnable task is going to get scheduled in that cycle anyway, it's not that significant from a scheduling perspective what order `decide_worker` gets called on them. If there were backpressure, and after some cutoff we didn't schedule any more, then priority order would matter a lot more, since it would decide which tasks make it in before the cutoff.
